### ISSUE TITLE
MPS Inf/Nan Loss

### ISSUE THREAD

**Issue by justusschock** [Lightning-AI/lightning#13285]
I am encountering a bug, namely some of my neural network's targets are corrupted using the M1 GPU. This does not happen on CPU. Specifically, some targets are set to large values (~ -2+e25), resulting in inf/nan loss.

I have isolated the behavior by stepping through the code and verifying at which steps the targets are still in the range (-1,1) as intended. The corrupted values first occur in the **batch** argument of **validation_step()**. I implemented a custom collate_fn to verify that nothing is wrong with my datasets and dataloaders. The correct targets leave the collate_fn, but then some of them appear corrupted within validation_step().

As I wrote above, the corrupted targets appear only when using the M1 GPU, whereas on the CPU everything works correctly. I could investigate this issue further if someone could tell me what happens between collate_fn and validation_step, i.e. where else I should step through the code to identify the source of corruption.

_Originally posted by @gloryVine in https://github.com/PyTorchLightning/pytorch-lightning/issues/13102#issuecomment-1154574207_

cc @akihironitta @justusschock

**Comment by justusschock** [comment]
Hi @gloryVine, I created a separate issue for this :)

Thanks for the report. To be honest, I likely suspect this to be an issue with core PyTorch and not our code. We are only doing the device mapping and as long as everything is running on MPS, we are doing our job correctly. 

As this is still an experimental feature (from both, PyTorch and our side) it might very well be the case, that some operations do not yet work as expected. For this it would be very helpful to pin down which operation specifically does not behave as expected. Therefore it would be helpful to run the network with the trainerflag `detect_anomaly=True` to get more insights on the operation. Once you do that, I would ask you to provide an example tensor for the operation to reproduce this behaviour to upstream it to pytorch :)

**Comment by carmocca** [comment]
@gloryVine Here is where the batch gets fetched (and `collate_fn` is called): https://github.com/PyTorchLightning/pytorch-lightning/blob/bfa8b7be2d99b980afa62f5cb0433326bcfd2ef0/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py#L127 and here is where `validation_step` gets called: https://github.com/PyTorchLightning/pytorch-lightning/blob/bfa8b7be2d99b980afa62f5cb0433326bcfd2ef0/pytorch_lightning/loops/epoch/evaluation_epoch_loop.py#L143 You can use a debugger to step around these calls.

**Comment by gloryVine** [comment]
@justusschock I ran the trainer with the flag, nothing changed, including the output.
@carmocca I did the following at the position in your first link: Immediately after `next(data_fetcher)`, I did the check `(abs(batch[1])>1000).any()`, which at some point is True (even though my targets are all supposed to be in the (-1,1) range). To make sure that it's not an issue with the data loader, **for the same batch** and with `y = data_fetcher.dataloader.dataset.y` I did the checks `(abs(y)>1000).any()` and also `abs(y.to("mps")>1000).any()`, both of which are `False`.
As written in my initial post, I also did the size check at the end of `collate_fn`, the targets are all ok there.

**Comment by LeonardoAlchieri** [comment]
I am experiencing a similar issue, when using a CUDA-trained model with MPS. What happens is that the model makes predictions during the training phase, but after `model.eval()`, all predictions are either 1s or NaNs. Any idea why?

**Comment by justusschock** [comment]
Hey @j0rd1smit thanks for the MWE and sorry for the inconvenience. I could indeed reproduce the issue on my side, but so far I have no clue what we could do causing overflows. I will investigate though and will get back to you once I found something!

cc @akihironitta @awaelchli who might have ideas where this could come from

**Comment by j0rd1smit** [comment]
I have been able to narrow it down to the `DataFetcher.move_to_device` ([link](https://github.com/Lightning-AI/lightning/blob/2415834aaf4f215ddb0d618bf415405d670f9114/src/pytorch_lightning/utilities/fetching.py#L289)).

In this function I noticed the following:
```python
def move_to_device(self, batch: Any) -> Any:
    if self.store_on_device:
        # Add this point no corrupted values
        batch = self.batch_to_device(batch)
        # Add this point there are corrupted values
    return batch
```
Interestingly, the issue resolves itself if I add an assert that checks for corrupted values like this:
```python
def move_to_device(self, batch: Any) -> Any:
    if self.store_on_device:
        batch = self.batch_to_device(batch)
        assert torch.sum(batch[1] >= 10) == 0
    return batch
```
I'm not sure why this works, but maybe it helps you to narrow down the problem.

**Comment by j0rd1smit** [comment]
I have been able to solve this issue for myself it by making the following change [here](https://github.com/Lightning-AI/lightning/blob/0b1a29b5eba423cec440ccc2761244f0d9efda4c/src/pytorch_lightning/utilities/apply_func.py#L346):
```python
_MPS_DEVICES = ("mps", torch.device("mps:0"))
...
if isinstance(data, Tensor) and device not in _CPU_DEVICES and device not in _MPS_DEVICES:
    kwargs["non_blocking"] = True
data_output = data.to(device, **kwargs)
```

I happy to make a PR for it. However, I wanted to discuss the solution first because I'm not if this solution solves the real problem or just one of the symptoms. What do you think @carmocca

**Comment by akihironitta** [comment]
@j0rd1smit Great finding! Quick search led me to https://github.com/pytorch/pytorch/issues/83015, and I've just confirmed that, with `.to(..., non_blocking=True)`, some elements of a tensor can somtimes get nan in pure PyTorch. I think this behaviour and the patch you suggest explain the stability difference between PL and PyTorch.

**Comment by justusschock** [comment]
@j0rd1smit That's indeed a great finding! the solution does look reasonable. Please go ahead with a PR!
